### PR TITLE
Update of the RPC migration to DD4hep 

### PR DESCRIPTION
**Pull request by slomeo**
#### PR description:

Update the RPC migration to DD4hep after @bsunanda  PRs #29842 #29943 and @ianna PR #29954.
 
#### PR validation:

1) validation made by Geometry/RPCGeometryBuilder/test/python/validateRPCGeometry_cfg.py

2) validation by runTheMatrix.py  -l 25202.1

25202.1_TTbar_13+TTbar_13+DIGIUP15APVSimu_PU25+RECOUP15_PU25+HARVESTUP15_PU25 Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED  - time date Fri May 29 17:36:47 2020-date Fri May 29 17:15:13 2020; exit: 0 0 0 0
1 1 1 1 tests passed, 0 0 0 0 failed

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
 nothing special

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by bsunanda** [comment]
We are trying to combine the access to ddd and dd4hep in MuonGeometryConstants. So in the RPCGeometryBuilder we can use the same MuonGeometryConstants for both ddd and dd4hep. Please utilize the 2 classes MuonGeometryConstants and MuonGeometryNumbering classes in all the builders as done for GEMGeometryBuilder and ME0GeometryBuilder

**Comment by slomeo** [comment]
@bsunanda : are you talking about PR #30043?

**Comment by bsunanda** [comment]
@slomeo Well in #30043 I only made the cff file for the ESProducer. But the use of MuonGeometryNumbering and MuonGeometryConstants for both ddd and dd4hep was ,ade in the earlier PR #29943

**Comment by slomeo** [comment]
@bsunanda : asap I'll update the code as you requested (i.e something similar to PR #29943 for GEMGeometryBuilderFromDDD)

**Comment by slomeo** [comment]
@bsunanda  I made the modifications that you suggested: the compilation is fine and the runTheMatrix test is fine, but my validation script now doesn't work anymore... please let me know what I have to modify

**Comment by slomeo** [comment]
@bsunanda : I fixed the problem with the validation script. Please @bsunanda @ianna @cvuosalo @civanch @kpedro88 @perrotta check again

**Comment by perrotta** [comment]
please test

File: Geometry/RPCGeometryBuilder/plugins/RPCGeometryESModule.cc [hunk 1]
```diff
@@ -33,14 +28,11 @@ RPCGeometryESModule::RPCGeometryESModule(const edm::ParameterSet& p)
       useDDD_{p.getUntrackedParameter<bool>("useDDD", true)},
       useDD4hep_{p.getUntrackedParameter<bool>("useDD4hep", false)} {
   auto cc = setWhatProduced(this);
-
   const edm::ESInputTag kEmptyTag;
-  if (useDDD_) {
+  if (useDDD_ || useDD4hep_) {
```
Comment (cvuosalo) [inline Geometry/RPCGeometryBuilder/plugins/RPCGeometryESModule.cc:32]: I think these lines have to be reverted. You can't use the old `DDCompactView` with DD4hep.

Comment (slomeo) [reply Geometry/RPCGeometryBuilder/plugins/RPCGeometryESModule.cc]: Done

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by slomeo** [comment]
This morning while I was starting to update the CSC migration to dd4hep, in a new IB, I got the same error occurred for the RPC (i.e some Chambers have a Z Position wrong). After a skype chat with @ianna I made a test updating the file DetectorConstruction/DDCMS/data/cms-geometry-2021.xml (for example the muonYoke.xml was v2 and not v3). The problem disappears. Because @ianna, in a next PR, will delete the file cms-geometry-2021.xml, I updated the file Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml and the file Configuration/​Geometry/​python/​GeometryDD4hepExtended2021_cff.py (instead of using cms-geometry-2021.xml) in order to fix the problem. This is the explanation for a new "commit -a".

**Comment by fabiocos** [comment]
please test

File: Configuration/Geometry/python/GeometryDD4hepExtended2021_cff.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 #
 # Ideal geometry, needed for simulation
 DDDetectorESProducer = cms.ESSource("DDDetectorESProducer",
-                                    confGeomXMLFiles = cms.FileInPath('DetectorDescription/DDCMS/data/cms-geometry-2021.xml'),
+                                    confGeomXMLFiles = cms.FileInPath('Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml'),
```
Comment (ianna) [inline Configuration/Geometry/python/GeometryDD4hepExtended2021_cff.py:8]: @slomeo - as we have discussed it this morning, please, remove this file from the PR

File: Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml [hunk 3]
```diff
@@ -4,286 +4,286 @@
   <close_geometry/>
 
   <IncludeSection>
-    <Include ref="Geometry/CMSCommonData/data/materials/2021/v1/materials.xml"/>
```
Comment (ianna) [inline Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml:7]: @slomeo - as we discussed it this morning, please remove this file from the PR. It's not correct as is.

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
 //
-// Author:  Sergio Lo Meo (sergio.lo.meo@cern.ch) following what Ianna Osburne made for DTs (DD4HEP migration)
+// Author:  Sergio Lo Meo (sergio.lo.meo@cern.ch) following what Ianna Osborne made for DTs (DD4HEP migration)
 //          Created:  Fri, 20 Sep 2019 
+//          Modified: Fri, 29 May 2020, following what Sunanda Banerjee made in PR #29842 PR #29943 and Ianna Osborne in PR #29954    
 */
-
-#include "Geometry/MuonNumbering/interface/DD4hep_RPCNumberingScheme.h"
-#include "Geometry/MuonNumbering/interface/DD4hep_MuonNumbering.h"
+//#include "Geometry/MuonNumbering/interface/DD4hep_RPCNumberingScheme.h"
```
Comment (ianna) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h:13]: please, remove commented out code

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h [hunk 5]
```diff
@@ -32,26 +31,25 @@ class RPCDetId;
 class RPCRoll;
 class MuonGeometryConstants;
 
-class RPCGeometryBuilderFromDDD {
+class RPCGeometryBuilder {
 public:
-  RPCGeometryBuilderFromDDD(bool comp11);
+  RPCGeometryBuilder(bool comp11);
```
Comment (ianna) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h:36]: please, remove `comp11` - we are already using C++17 :-)

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h [hunk 6]
```diff
@@ -23,7 +22,7 @@ class DDFilteredView;
 namespace cms {
   class DDFilteredView;
   class DDCompactView;
-  class MuonNumbering;
+  //class MuonNumbering;
```
Comment (ianna) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h:25]: please, remove commented out code

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h [hunk 7]
```diff
@@ -32,26 +31,25 @@ class RPCDetId;
...
 class MuonGeometryConstants;
 
-class RPCGeometryBuilderFromDDD {
+class RPCGeometryBuilder {
 public:
-  RPCGeometryBuilderFromDDD(bool comp11);
+  RPCGeometryBuilder(bool comp11);
 
-  ~RPCGeometryBuilderFromDDD();
+  ~RPCGeometryBuilder();
```
Comment (ianna) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h:38]: in C++11 and above you do not need to write an empty destructor. Please, remove it.

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc [hunk 8]
```diff
@@ -4,55 +4,47 @@
...
 #include "DataFormats/Math/interface/CMSUnits.h"
 #include "DataFormats/Math/interface/GeantUnits.h"
 
 using namespace cms_units::operators;
 
-RPCGeometryBuilderFromDDD::RPCGeometryBuilderFromDDD(bool comp11) : theComp11Flag(comp11) {}
+RPCGeometryBuilder::RPCGeometryBuilder(bool comp11) : theComp11Flag(comp11) {}
 
-RPCGeometryBuilderFromDDD::~RPCGeometryBuilderFromDDD() {}
+RPCGeometryBuilder::~RPCGeometryBuilder() {}
```
Comment (ianna) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc:36]: in C++11 and above you do not need to write an empty destructor. Please, remove it.

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc [hunk 9]
```diff
@@ -4,55 +4,47 @@
...
 #include "DetectorDescription/DDCMS/interface/DDSpecParRegistry.h"
-#include "Geometry/MuonNumbering/interface/DD4hep_RPCNumberingScheme.h"
-
 #include "DataFormats/Math/interface/CMSUnits.h"
 #include "DataFormats/Math/interface/GeantUnits.h"
 
 using namespace cms_units::operators;
 
-RPCGeometryBuilderFromDDD::RPCGeometryBuilderFromDDD(bool comp11) : theComp11Flag(comp11) {}
+RPCGeometryBuilder::RPCGeometryBuilder(bool comp11) : theComp11Flag(comp11) {}
```
Comment (ianna) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc:34]: please, remove `comp11` - we use C++17 already

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc [hunk 10]
```diff
@@ -251,19 +241,15 @@ RPCGeometry* RPCGeometryBuilderFromDDD::buildGeometry(DDFilteredView& fview,
 }
 
 // for DD4hep
-RPCGeometry* RPCGeometryBuilderFromDDD::buildGeometry(cms::DDFilteredView& fview,
-                                                      const cms::MuonNumbering& muonConstants) {
+RPCGeometry* RPCGeometryBuilder::buildGeometry(cms::DDFilteredView& fview, const MuonGeometryConstants& muonConstants) {
   RPCGeometry* geometry = new RPCGeometry();
```
Comment (ianna) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc:245]: please, use `unique_ptr`

**Review by ianna — CHANGES_REQUESTED** [review]
@slomeo - please, make the `RPGGeometry*` a smart pointer

File: Configuration/Geometry/python/GeometryDD4hepExtended2021_cff.py (hunk 2 above)
Comment (slomeo) [reply Configuration/Geometry/python/GeometryDD4hepExtended2021_cff.py:8]: done

File: Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml (hunk 3 above)
Comment (slomeo) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml:7]: done

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h (hunk 4 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h]: done

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h (hunk 5 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h]: done

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h (hunk 6 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h]: done

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h (hunk 7 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.h]: done

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc (hunk 8 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc]: done

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc (hunk 9 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc]: done

File: Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc (hunk 10 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryBuilder.cc]: done

**Comment by perrotta** [comment]
please test

**Comment by slomeo** [comment]
@ianna @cvuosalo looking at https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3104a3/6802/summary.html it seems that there is a problem related to cms-geometry-2021.xml. I deleted this file so I think it is necessary to update some scripts. Am I correct? If I'm correct, can you please take care of it?

**Comment by ianna** [comment]
> @ianna @cvuosalo looking at https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3104a3/6802/summary.html it seems that there is a problem related to cms-geometry-2021.xml. I deleted this file so I think it is necessary to update some scripts. Am I correct? If I'm correct, can you please take care of it?

@slomeo - what I was asking is to remove your changes to  `Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml` and `Configuration/Geometry/python/GeometryDD4hepExtended2021_cff.py` in this PR.

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by slomeo** [comment]
@ianna do you  mean a "git restore"?

**Comment by kpedro88** [comment]
@slomeo 'git restore' is not a command. You should 'git revert' the commit in which you removed the file, additionally undo any changes you previously made to the files mentioned by @ianna, and then squash the commits in this PR to avoid churn in the history and blame.

**Comment by slomeo** [comment]
@ianna @kpedro88 I think it's better to re-start with a new PR in a clean IB. I'll close this PR and I'll make a new PR asap.

**Comment by slomeo** [comment]
I made a new PR, please look to #30121

**Pull request closed without merging**